doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,752 @@
+support for sockets operating in an "unconnected" state. Currently, sending a
+UDP packet from a dgram `BIO` using an unconnected socket requires the caller to
+first use `BIO_ctrl()` to set the destination address and then call `BIO_write()`
+to actually send the packet. A similar process is required for obtaining the peer
+address from a received packet. This multi-call approach is unlikely to scale
+well in a multi threaded application.
+
+A further problem is the existing API provides no mechanism for sending or
+receiving multiple messages at the same time. Operating systems provide various
+mechanisms for optimising sending/receiving data if this can be achieved. Note
+that OpenSSL does not need to implement support for this in its own `BIO`s in its
+initial releases - but adding support for it in the API enables others to do so,
+and enables it to be added easily to OpenSSL's own `BIO`s at any point.
+
+The existing API also provides no support for setting/obtaining the local
+address for a message to be sent/received. This may be useful in the case where
+a socket is bound to multiple address using `INADDR_ANY` or `IN6ADDR_ANY`.
+
+To solve all of the above issues two new functions are proposed: `BIO_sendmmsg()`
+and `BIO_recvmmsg()`. There will also be associated `BIO_METHOD` functions to set
+implementations for these. Support can be added to fall back to the existing
+sending/receiving approach in the event that a `BIO` does not support the new
+`BIO_METHOD` functions.
+
+A demonstration client using the new SSL API functions has been provided in
+[Appendix B](#appendix-b).
+
+# <a id='appendix-a'></a>Appendix A: SSL API ssl.h and bio.h additions
+
+Proposed additions to the ssl.h header file are as follows:
+
+
+````
+/* Methods representing the QUIC protocol */
+__owur const SSL_METHOD *QUIC_method(void);
+__owur const SSL_METHOD *QUIC_server_method(void);
+__owur const SSL_METHOD *QUIC_client_method(void);
+
+/*
+ * An SSL object represents a single stream. In QUIC multiple SSL objects can be
+ * associated with the same connection id. In (D)TLS this would just generate
+ * a unique number for each SSL object.
+ */
+uint64_t SSL_get_conn_id(SSL *ssl);
+
+/* Object to represent various events that might occur */
+typedef struct ssl_event_st SSL_EVENT;
+
+/*
+ * Object to manage a collection of event sources. SSL_EVENT_CTX objects are
+ * ref counted.
+ */
+typedef struct ssl_event_ctx_st SSL_EVENT_CTX;
+
+/* Set the SSL_EVENT_CTX inherited by SSL objects created from this SSL_CTX */
+int SSL_CTX_set1_SSL_EVENT_CTX(SSL_CTX *ctx, SSL_EVENT_CTX *ectx);
+
+/*
+ * Override the SSL_EVENT_CTX to use for an inidivudal SSL object. Calling
+ * SSL_set_bio(), SSL_set0_rbio() or SSL_set0_wbio() on an SSL object will
+ * replace any existing SSL_EVENT_CTX with a new one containing those BIOs.
+ * The SSL_CTX used to create the SSL object and SSL_EVENT_CTX object must use
+ * the same SSL_METHOD.
```

Comment:
@hlandau wrote:

> Can you elaborate on this? This sounds significant/surprising. An application probably doesn't expect to have an SSL object suddenly move to another SSL_EVENT_CTX. Maybe we should just change this to be unsupported and if they want to do this, they can create a new SSL_EVENT_CTX and set it on the SSL object themselves?

We've separated out the concept of `SSL_EVENT_CTX` from the `SSL_CTX`. I envisaged that you might want to use a single `SSL_CTX` but have multiple `SSL_EVENT_CTX` objects used. Consider a client creating two connections to two different servers. We might use the same `SSL_CTX` for both, but it seems reasonable to have 2 different `SSL_EVENT_CTX`s.

I envisaged that you might create an `SSL` object which would inherit a "default" `SSL_EVENT_CTX` from the `SSL_CTX` but then you want to override that with a different one. Probably moving to a different `SSL_EVENT_CTX` *after* the handshake has started is probably unwise so we could possibly prohibit that.